### ISSUE TITLE
 Error: Class 'Shrink\Lib\ShrinkCompiler\ShrinkCompiler' not found 

### ISSUE THREAD

**Issue by danielignatius** [trentrichardson/cakephp-shrink#7]
I've followed the instruction below, 

bootstrap.php
Plugin::load('Shrink',['autoload'=>true]);

AppController.php
$helpers = array('Form','Html',Shrink.Shrink');

pp/plugin
Shrink folder

but I not sure why am I getting a  Error: Class 'Shrink\Lib\ShrinkCompiler\ShrinkCompiler' not found in File D:\home\site\wwwroot\plugins\Shrink\src\Lib\ShrinkType.php.

I've added the class **use Shrink\Lib\ShrinkCompiler\ShrinkCompiler; ** in ShrinkType.php. it still showing the error.

Any advise please?

**Comment by trentrichardson** [comment]
When including the file via the helper are you including the file
extension? Also what is the file extension?

**Comment by danielignatius** [comment]
Thanks Richardson for your quick reply. I would like to compress my all my css files into one css file. What am I missing still?

**AppController.php**
public $helpers = ['Form','Html','Time','Shrink.Shrink'];

**Layout**
$this->Shrink->css(['/css/uni-icons.css','/assets/css/style.css']);
echo $this->Shrink->fetch('css');

**Comment by trentrichardson** [comment]
That looks correct.. are you using composer, and is it in composer.json?
Maybe try composer update.

**Comment by trentrichardson** [comment]
Looking back at the error it looks like it is searching the plug-ins
directory for the code and not via composer. If you're familiar with
composer try installing with it as it knows how to map out and lookup
namespaces

**Comment by danielignatius** [comment]
I've manually downloaded the plugin and place it in my plugin folder and set it as Plugin::load('Shrink',['autoload'=>true]); in the bootstrap.php file. I am not very familiar how to work with install a plugin using a composer. Therefore, I've manually added it into my application folder.

**Comment by danielignatius** [comment]
But I pretty sure that the plugins and the files is in the correct directory as required.

**Comment by trentrichardson** [comment]
Try this "Shrink/Shrink" to define the top level namespace in bootstrap.php:

http://book.cakephp.org/3.0/en/plugins.html#loading-a-plugin

**Comment by danielignatius** [comment]
I've tried both namespace Shrink\Shrink; or use Shrink\Shrink; in bootstrap.php: Both doesn't work.

**Comment by trentrichardson** [comment]
Tried doing just this in bootstrap (note the forward slash, not backward
slash)?

Plugin::load('Shrink/Shrink', ['autoload'=>true]);

Or have you tried:

Plugin::loadAll();

Sorry I'm sure you've already gone through a lot of this..

**Comment by danielignatius** [comment]
Plugin::load('Shrink/Shrink', ['autoload'=>true]); - WILL GIVE Make sure your plugin Shrink/Shrink is in the D:\home\site\wwwroot\plugins\ directory and was loaded. As the folder is called Shrink.

**Comment by danielignatius** [comment]
As for Plugin::loadAll(); it gives the below

Error: Shrink.ShrinkHelper could not be found.

Make sure your plugin was loaded from config\bootstrap.php and Composer is able to autoload its classes, see Loading a plugin and Plugins - autoloading plugin classes

**Comment by trentrichardson** [comment]
And you did rename the folder to "Shrink" inside the plugin folder?

**Comment by danielignatius** [comment]
Yes, the folder is called 'Shrink'. previously when i downloaded it was 'cakephp-shrink-master'

**Comment by danielignatius** [comment]
Should the folder called Shrink.Shrink?

**Comment by trentrichardson** [comment]
"Shrink" should be the correct folder name. Sorry I'm not at a computer to
help debug...

**Comment by danielignatius** [comment]
That's okay. You've been helpful. :)

**Comment by danielignatius** [comment]
I've been trying solving this since last night. 

I thought was a missing class and in the ShrinkType.php I have added use 'Shrink\Lib\ShrinkCompiler\ShrinkCompiler'; But it still not okay. =(

===
namespace Shrink\Lib;

**use 'Shrink\Lib\ShrinkCompiler\ShrinkCompiler' ;**

class ShrinkType{

	/**
	* Create a new instance of the requested Shrink Compiler class
	* @param string $type - 'js', 'css', 'less'. This should be the input type
	* @param array $settings - passthrough of complete settings from helper
	* @return class - new instance of the requested type
	*/
	static function getCompiler($type, $settings){

        $class = __NAMESPACE__ . '\\ShrinkCompiler\\' .
            'ShrinkCompiler'. strtoupper(substr($type,0,1)).strtolower(substr($type,1));
        return new $class($settings);
	}

**Comment by trentrichardson** [comment]
It may be tomorrow sometime, but when I get to my computer and can help
debug it a little more.

I think it boils down to that Shrink namespace not being registered so that
php knows where to find it. With cake 3 I've always used composer and I
know it handles this automatically. The plugin may need to be tweaked for
this.

And just for reference you are running php 5.4+?

**Comment by danielignatius** [comment]
Sure. Thanks so much. I am running php 5.6.7.

**Comment by danielignatius** [comment]
Hi Trent any lucks with the Shrink namespace? I really like to use your plugins. Thanks!

**Comment by trentrichardson** [comment]
I modified an existing cake app to manually place the plugin into the plugins folder.  I didn't get your exact error but I did notice a similar error that a class was not found.  I think this is because php can't find the "Shrink" namespace.  I played around with it and found that if I simply change in composer.json from:

```json
//...
    "autoload": {
        "psr-4": {
            "App\\": "src"
        }
    },
//...
```

to include an entry for Shrink like so:

```json
//...
    "autoload": {
        "psr-4": {
            "App\\": "src",
            "Shrink\\": "./plugins/Shrink/src"
        }
    },
//...
```
and simply run `composer update` or `composer dumpautoload` that the error went away.  My bootstrap had `Plugin::load('Shrink');` and AppController.php included the helper with `public $helpers = ['Shrink.Shrink'];`.

Simply put, I think composer handles [spl_autoload_register()](http://php.net/manual/en/function.spl-autoload-register.php) which is what allows php to lookup namespaces in a customized way.  When not using composer these namespaces don't get mapped and classes can't be found by namespaces.  

I know its a small learning curve but I would highly recommend trying to transition to using Composer since CakePHP 3 is so tighly coupled to its new design.   If nothing more than to install composer and run `composer update` to build the autoload.  I can share with you my workflow since moving to cake 3 and composer if you like. Since making the switch I can honestly say I won't be going back as it was a huge improvement.  It cleaned up my projects and made deployment and testing easier.

You may be able to simply download composer.phar into your project and run it like:
```sh
php composer.phar dumpautoload
```

**Comment by trentrichardson** [comment]
Sorry, it also just dawned on me that the Shrink plugin itself requires composer to load it's dependencies like CSSMin.. I know that doesn't help matters much..

https://github.com/trentrichardson/cakephp-shrink/blob/master/composer.json

So it looks like the current state of Shrink does require composer to work properly..

**Comment by danielignatius** [comment]
Trust me, I am all willing to learn. I am really happy and appreciate if you can share your workflow with me. For instance, how do I add a plugin(like your plugin or some others) or even updating the latest CakePHP Core using composer. I find it hard to follow when the "baby steps" are missing from the book guide. So far, I learn a lot from you. I am using a Mac. First, I will download the composer.phar. And can you provide me with the list of command or workflow I need to do to download your plugin using composer. Really appreciate your help.

**Comment by trentrichardson** [comment]
I don't mind at all sharing my workflow. Send me your email and I will send
it over to you.

**Comment by danielignatius** [comment]
Hey Trent, my email is danielignatius@me.com. Thanks! I really hope I get your plugin working!!

**Comment by danielignatius** [comment]
I've installed the composer.phar. on the root of my application. So is it true that for each app I need to install one composer.phar?

<img src="http://s2.postimg.org/43dyewutl/Screen_Shot_2016_02_02_at_3_39_31_PM.png"/>

**Comment by trentrichardson** [comment]
Ideally you have a single composer that is system wide. I was more or less saying previously that you _could_ download it into your project if you needed something very simple..

I'm going to send you an email with more info, and close this thread for now.

**Comment by EgoLord** [comment]
Hi,
just bumped in the same error.

This is the steps I've followed:

1) updated my composer.json

```
"require": {
        //lot of requirements..
        "trentrichardson/cakephp-shrink": "~3.0"
    },
```

2) added this line in bootstrap.php
`Plugin::load('Shrink', ['ignoreMissing' => true, 'bootstrap' => true, 'routes' => true, 'autoload' => true]);`

3) added this into AppView.php inside initialize func
```
$this->loadHelper('Shrink.Shrink', [
		    'js'=>[
		            'path'=>'js/',        // folder to find src js files
		            'cachePath'=>'js/',   // folder to create cache files
		            'minifier'=>'jsmin'   // minifier to minify, false to leave as is
		    ],
		    'css'=>[
		            'path'=>'css/',       // folder to find src css files
		            'cachePath'=>'css/',  // folder to create cache files
		            'minifier'=>'cssmin', // minifier name to minify, false to leave as is
		            'charset'=>'utf-8'    // charset to use
		    ],
		    'url'=>'',                     // url without ending /, incase you access from another domain
		    'prefix'=>'shrink_',           // prefix the beginning of cache files
		    'debugLevel'=>1                // compared against Core.debug, eq will recompile, > will not minify
		]);
```

4) Updated my layout with:
`<?= $this->Shrink->css($css_basic, true) ?>`


Did I miss something?

**Comment by EgoLord** [comment]
Ok it was a little bubu about the plugin assets

in /vendor/trentrichardson/Shrink/src/Lib/Shrink.php on line 104 I changed the follow foreach:

```
foreach($files as $k=>$v){
	$tmpf = new File(preg_replace('/(\/+|\\+)/', DS, WWW_ROOT .DS. ($v[0]=='/'? '':$this->settings[$type]['path']) .DS. $v));
	$files[$k] = [
			'file'=>$tmpf,
			'rel_path'=>$v
		];
	$srcMod = $tmpf->lastChange();
	if($srcMod > $maxAge){
		$maxAge = $srcMod;
	}
}
```

With this one:

```
foreach($files as $k=>$v){
	$webroot = WWW_ROOT;
	
	//Retrieve plugin
	list($plugin, $v) = pluginSplit($v);
	
	//Change webroot accordigly to plugin if present
	if(!is_null($plugin))
		$webroot = realpath(WWW_ROOT . '..' . DS . 'plugins' . DS . $plugin . DS . 'webroot');
	
	//Add realpath to aim a real directory
	$tmpf_path = realpath(preg_replace('/(\/+|\\+)/', DS, $webroot .DS. ($v[0]=='/'? '':$this->settings[$type]['path']) .DS. $v));
	
	$tmpf = new File($tmpf_path);
	$files[$k] = [
			'file'=>$tmpf,
			'rel_path'=>$v
		];
	$srcMod = $tmpf->lastChange();
	if($srcMod > $maxAge){
		$maxAge = $srcMod;
	}
}

```

It should fix the problem when you want to shrink a file from a plugin and the file is not into the default folder (ex. folder assets on the same level of css and js folders).

So when you have a css/js list like the following Shrink could find the file correctly:
```
$css_basic = [
		$PLUGIN . '.' . '../assets/uniform/css/uniform.default.min.css',
		$PLUGIN . '.' . '../assets/bootstrap/css/bootstrap.min.css',
		//Lots of css files
	];
echo $this->Shrink->css($css_basic, true);
```

**Comment by riddlemd** [comment]
Still not fixed hu? I just ran into this same issue.

**Comment by trentrichardson** [comment]
Thanks @riddlemd. I applied this fix. Not sure if you were using master branch or cakephp-v3.0 branch, but beware there have been changes since 2 years ago.
